### ISSUE TITLE
Future of matrix mixin classes

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#25601]
I'm not sure they are used now,
and it seems like those tests start to confuse new contributors

https://github.com/sympy/sympy/blob/5ea23d3483d8546d60494aba3e4363e3271d378c/sympy/matrices/tests/test_commonmatrix.py#L28-L82

Although they were public classes and intended by the original authors,
I am skeptical if they are easy to use or compose well now.
If it is just from end user app, or SymPy deprecates things like that in major, I'd like to remove them.
However, it is not easy to deprecate things from SymPy, so we'd need to be careful from here.

- We can remove them from the tests because they can duplicate the tests.
- We can deprecate them because they are not working properly any more or they make things more harder for domainmatrix translation or such.

**Comment by asmeurer** [comment]
I never really got the point of them. The person who added them had some reasons for doing so (you'll have to look at the old PRs). I would say the only real reason to do something like this is if it makes the code simpler. If it doesn't, then we should get rid of it. 

IMO, if we want to split out matrices into different classes, it should only be done

1. By internal representation (like sparse vs. dense), and/or
2. using actual mathematical distinctions. There's no such mathematical thing as a `MatrixDeterminant` or a `MatrixReductions`. If we want to think about matrices that do or don't support different structures, we should be thinking about that in terms of actual algebraic properties, e.g., matrices over a ring vs. matrices over a field. 

> We can deprecate them in some ways like vendoring the old source code, and isolate from giving updates.

I don't think that would work (the classes would be distinct from the actual Matrix class). But even so I don't think it's necessary. Half the classes already start with an underscore, meaning they are definitely private.

**Comment by sylee957** [comment]
I don't think that it is very difficult to remove them from test cases, however, it can get a refutation about why do it because there is not much problems of keeping them in SymPy test cases yet, to catch some issues.
I think that having nontrivial test cases like using Hypothesis should replace them

**Comment by oscarbenjamin** [comment]
> there is not much problems of keeping them in SymPy test cases yet, to catch some issues.

I don't think that they do catch any real issues and I have seen many contributors trying to make improvements to Matrix stumble over the need to write code that can also work for things like `ShapingOnlyMatrix`. There are all of these mixins and tests that seem designed to check that they can be inherited to create new matrix classes but I don't think it is really possible to do anything useful with them beyond the only thing that SymPy actually uses them for: they provide the methods for the Matrix class.

In general inheritance is usually a bad design. The mixins could be a nice way to break up the code for the Matrix class but testing explicitly that they can be subclassed to create other classes is a bad idea because it is not worthwhile to create those other classes. Also we don't need many different kinds of Matrix class. Users generally just want to use Matrix or perhaps DomainMatrix.

Composition and just ordinary functions are almost always a better way to handle code-sharing rather than inheritance. DomainMatrix uses composition rather than inheritance and can internally hold either a DDM, SDM or DFM each of which is a different data structure with specialised algorithms for low-level interaction with the data structure. In this design the idea is that any code that could be shared between DDM, SDM and DFM can just be located in DomainMatrix instead or we can convert representations like `.to_ddm()` if an algorithm is implemented for one representation but not another. That is a much better way of doing code-sharing and does not use any inheritance meaning that each method in each class applies precisely to instances of that class without needing to worry about any subclasses overriding things.

**Comment by oscarbenjamin** [comment]
After #25911 these classes are all separated from Matrix itself but are still left in position so that they can be deprecated.
